File: app/reporter/markdown.py
"""Markdown-to-HTML conversion and confidence-token highlighting.

Provides functions for converting AI-produced Markdown (headings, lists, bold,
italic, code spans, tables, fenced code blocks) into safe HTML suitable for
embedding in forensic reports.  Also includes Jinja2 filter wrappers for
direct use in templates.

Key capabilities:

* **Inline formatting** -- bold, italic, backtick code spans.
* **Block elements** -- headings, ordered/unordered lists, fenced code blocks,
  paragraphs with ``<br>`` line breaks.
* **Tables** -- pipe-delimited Markdown tables rendered as ``<table>`` elements.
* **Confidence highlighting** -- explicit confidence labels
  (``Confidence: HIGH`` etc.) wrapped in coloured ``<span>`` elements.

Attributes:
    CONFIDENCE_PATTERN: Regex matching explicit confidence labels.
    CONFIDENCE_CLASS_MAP: Maps severity label strings to CSS class names.
    MARKDOWN_HEADING_PATTERN: Regex matching ATX-style headings (``#``--``######``).
    MARKDOWN_ORDERED_LIST_PATTERN: Regex matching ordered list items.
    MARKDOWN_UNORDERED_LIST_PATTERN: Regex matching unordered list items.
    MARKDOWN_BOLD_STAR_PATTERN: Regex matching ``**bold**`` syntax.
    MARKDOWN_BOLD_UNDERSCORE_PATTERN: Regex matching ``__bold__`` syntax.
    MARKDOWN_ITALIC_STAR_PATTERN: Regex matching ``*italic*`` syntax.
    MARKDOWN_ITALIC_UNDERSCORE_PATTERN: Regex matching ``_italic_`` syntax.
    MARKDOWN_TABLE_SEPARATOR_CELL_PATTERN: Regex matching table separator cells.
"""

from __future__ import annotations

from collections.abc import Sequence
import html
import re
from typing import Any

from markupsafe import Markup, escape

from ..utils import stringify as _stringify_canonical

__all__ = [
    "CONFIDENCE_CLASS_MAP",
    "CONFIDENCE_PATTERN",
    "format_block",
    "format_markdown_block",
    "highlight_confidence_tokens",
    "markdown_to_html",
    "render_inline_markdown",
]

CONFIDENCE_PATTERN = re.compile(
    r"\bconfidence\b(?:\s+level)?\s*(?::|=|-|\bis\b)?\s*"
    r"(?:\*\*|__|\*|_|`|<strong>|<em>|<code>)*"
    r"(CRITICAL|HIGH|MEDIUM|LOW)(?![A-Za-z0-9])"
    r"(?:\*\*|__|\*|_|`|</strong>|</em>|</code>)*",
    re.IGNORECASE,
)
MARKDOWN_HEADING_PATTERN = re.compile(r"^(#{1,6})\s+(.*)$")
MARKDOWN_ORDERED_LIST_PATTERN = re.compile(r"^\d+\.\s+(.*)$")
MARKDOWN_UNORDERED_LIST_PATTERN = re.compile(r"^[-*]\s+(.*)$")
MARKDOWN_BOLD_STAR_PATTERN = re.compile(r"\*\*(.+?)\*\*")
MARKDOWN_BOLD_UNDERSCORE_PATTERN = re.compile(
    r"(^|\s)__([^\s_](?:.*?[^\s_])?)__(?=$|\s|[.,;!?:)])"
)
MARKDOWN_ITALIC_STAR_PATTERN = re.compile(r"(?<!\*)\*(?!\*)(.+?)(?<!\*)\*(?!\*)")
MARKDOWN_ITALIC_UNDERSCORE_PATTERN = re.compile(
    r"(^|\s)_([^\s_](?:.*?[^\s_])?)_(?=$|\s|[.,;!?:)])"
)
MARKDOWN_TABLE_SEPARATOR_CELL_PATTERN = re.compile(r"^:?-{3,}:?$")

CONFIDENCE_CLASS_MAP = {
    "CRITICAL": "confidence-critical",
    "HIGH": "confidence-high",
    "MEDIUM": "confidence-medium",
    "LOW": "confidence-low",
}


def _stringify(value: Any, default: str = "") -> str:
    """Convert *value* to a stripped string, returning *default* if empty.

    Delegates to the canonical :func:`app.utils.stringify` implementation.

    Args:
        value: Any value to convert to string.
        default: Fallback when *value* is None or empty after stripping.

    Returns:
        The stripped string representation, or *default*.
    """
    return _stringify_canonical(value, default)


def highlight_confidence_tokens(text: str) -> str:
    """Wrap confidence-related severity tokens in coloured ``<span>`` elements.

    Matches severity keywords only in explicit confidence contexts, not as
    ordinary English adjectives or all-caps prose.  Examples include
    ``"Confidence: HIGH"``, ``"confidence level: medium"``, and
    ``"confidence high"``.

    Args:
        text: Pre-escaped HTML string to scan for severity tokens.

    Returns:
        The input string with confidence severity tokens wrapped in spans.
    """
    def _replace_confidence(match: re.Match[str]) -> str:
        """Replace a confidence token match with a styled span.

        Only the keyword portion is wrapped in the coloured span; the
        confidence prefix text is preserved as-is.
        """
        keyword = match.group(1)
        label = keyword.upper()
        css_class = CONFIDENCE_CLASS_MAP.get(label, "confidence-unknown")
        span = f'<span class="confidence-inline {css_class}">{label}</span>'

        full = match.group(0)
        keyword_start = full.lower().rfind(keyword.lower())
        prefix = full[:keyword_start]
        suffix = full[keyword_start + len(keyword):]
        return f'{prefix}{span}{suffix}'

    return CONFIDENCE_PATTERN.sub(_replace_confidence, text)


def render_inline_markdown(value: str, *, escape_html: bool = True) -> str:
    """Render inline Markdown formatting to HTML.

    Handles backtick code spans, bold (``**`` and ``__``), italic
    (``*`` and ``_``), and confidence-token highlighting.

    .. warning:: Security consideration

        This function escapes HTML by default.  Internal callers that
        already pass pre-escaped text use ``escape_html=False`` to avoid
        double escaping.

    Args:
        value: Inline Markdown text.
        escape_html: When ``True``, apply :func:`html.escape` to *value*
            before processing Markdown formatting.  Defaults to ``True``.

    Returns:
        An HTML string with inline formatting applied.
    """
    source = str(value or "")
    if escape_html:
        source = html.escape(source)
    if not source:
        return ""

    parts = re.split(r"(`[^`\n]*`)", source)
    output: list[str] = []
    for part in parts:
        if not part:
            continue
        if len(part) >= 2 and part.startswith("`") and part.endswith("`"):
            output.append(f"<code>{part[1:-1]}</code>")
            continue

        escaped = part
        escaped = MARKDOWN_BOLD_STAR_PATTERN.sub(r"<strong>\1</strong>", escaped)
        escaped = MARKDOWN_BOLD_UNDERSCORE_PATTERN.sub(r"\1<strong>\2</strong>", escaped)
        escaped = MARKDOWN_ITALIC_STAR_PATTERN.sub(r"<em>\1</em>", escaped)
        escaped = MARKDOWN_ITALIC_UNDERSCORE_PATTERN.sub(r"\1<em>\2</em>", escaped)
        escaped = highlight_confidence_tokens(escaped)
        output.append(escaped)
    return "".join(output)


def _split_table_row(value: str) -> list[str]:
    """Split a Markdown table row into cell strings.

    Strips leading and trailing pipe characters before splitting on
    the remaining pipes.  Returns an empty list when *value* does
    not contain a pipe.

    Args:
        value: A single Markdown table row line.

    Returns:
        A list of stripped cell strings, or an empty list.
    """
    row_text = str(value or "")
    if "|" not in row_text:
        return []

    trimmed = row_text.strip()
    if not trimmed or "|" not in trimmed:
        return []

    if trimmed.startswith("|"):
        trimmed = trimmed[1:]
    if trimmed.endswith("|"):
        trimmed = trimmed[:-1]

    return [cell.strip() for cell in trimmed.split("|")]


def _is_table_separator_row(cells: Sequence[str]) -> bool:
    """Determine whether *cells* represent a Markdown table separator row.

    A separator row consists entirely of cells matching the pattern
    ``:?-+:?`` (e.g. ``---``, ``:---:``, ``---:``).

    Args:
        cells: List of cell strings from a split table row.

    Returns:
        *True* when every cell matches the separator pattern.
    """
    if not cells:
        return False
    return all(MARKDOWN_TABLE_SEPARATOR_CELL_PATTERN.match(str(cell).strip()) for cell in cells)


def _normalize_table_row_cells(cells: Sequence[str], expected_count: int) -> list[str]:
    """Pad or truncate *cells* to exactly *expected_count* entries.

    Cells beyond *expected_count* are discarded; missing cells are
    filled with empty strings.

    Args:
        cells: Raw cell values from a split table row.
        expected_count: The desired number of columns.

    Returns:
        A list of exactly *expected_count* stripped cell strings.
    """
    normalized = [str(cell).strip() for cell in cells[:expected_count]]
    if len(normalized) < expected_count:
        normalized.extend([""] * (expected_count - len(normalized)))
    return normalized


def _render_table_html(header_cells: Sequence[str], body_rows: Sequence[Sequence[str]]) -> str:
    """Render a parsed Markdown table as an HTML ``<table>`` element.

    Each cell value is processed through :func:`render_inline_markdown`
    so that inline formatting (bold, italic, code spans) is preserved.

    Args:
        header_cells: List of header cell strings.
        body_rows: List of body row lists, each containing cell
            strings matching the header column count.

    Returns:
        An HTML string containing the complete ``<table>`` element.
    """
    header_html = "".join(f"<th>{render_inline_markdown(cell, escape_html=False)}</th>" for cell in header_cells)
    table_html = [f"<table><thead><tr>{header_html}</tr></thead>"]

    if body_rows:
        rows_html: list[str] = []
        for row in body_rows:
            row_html = "".join(f"<td>{render_inline_markdown(cell, escape_html=False)}</td>" for cell in row)
            rows_html.append(f"<tr>{row_html}</tr>")
        table_html.append(f"<tbody>{''.join(rows_html)}</tbody>")

    table_html.append("</table>")
    return "".join(table_html)


def markdown_to_html(value: str) -> str:
    """Convert a complete Markdown text block to HTML.

    Supports headings (``#`` through ``######``), ordered and
    unordered lists, fenced code blocks (triple backticks), tables,
    inline formatting (bold, italic, code spans), and
    confidence-token highlighting.  Paragraphs are wrapped in
    ``<p>`` tags with ``<br>`` line breaks.

    Args:
        value: Raw Markdown text (may contain multiple blocks).

    Returns:
        An HTML string with all recognised Markdown constructs
        converted to their HTML equivalents.
    """
    value = html.escape(str(value))
    lines = value.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    blocks: list[str] = []
    paragraph_lines: list[str] = []
    list_items: list[str] = []
    list_type = ""
    in_code_fence = False
    code_lines: list[str] = []

    def flush_paragraph() -> None:
        """Flush accumulated paragraph lines into a ``<p>`` block."""
        nonlocal paragraph_lines
        if not paragraph_lines:
            return
        paragraph_text = "\n".join(paragraph_lines)
        rendered = render_inline_markdown(paragraph_text, escape_html=False).replace("\n", "<br>\n")
        blocks.append(f"<p>{rendered}</p>")
        paragraph_lines = []

    def flush_list() -> None:
        """Flush accumulated list items into an ``<ol>`` or ``<ul>`` block."""
        nonlocal list_items, list_type
        if not list_items or not list_type:
            list_items = []
            list_type = ""
            return
        items_html = "".join(f"<li>{item}</li>" for item in list_items)
        blocks.append(f"<{list_type}>{items_html}</{list_type}>")
        list_items = []
        list_type = ""

    def flush_code_fence() -> None:
        """Flush accumulated code lines into a ``<pre><code>`` block."""
        nonlocal code_lines
        code_text = "\n".join(code_lines)
        blocks.append(f"<pre><code>{code_text}</code></pre>")
        code_lines = []

    index = 0
    while index < len(lines):
        line = lines[index]
        stripped = line.strip()

        if in_code_fence:
            if stripped.startswith("```"):
                in_code_fence = False
                flush_code_fence()
            else:
                code_lines.append(line)
            index += 1
            continue

        if stripped.startswith("```"):
            flush_paragraph()
            flush_list()
            in_code_fence = True
            code_lines = []
            index += 1
            continue

        if not stripped:
            flush_paragraph()
            flush_list()
            index += 1
            continue

        header_cells = _split_table_row(line)
        if header_cells and index + 1 < len(lines):
            separator_cells = _split_table_row(lines[index + 1])
            if (
                separator_cells
                and len(header_cells) == len(separator_cells)
                and _is_table_separator_row(separator_cells)
            ):
                flush_paragraph()
                flush_list()

                expected_columns = len(header_cells)
                normalized_header = _normalize_table_row_cells(header_cells, expected_columns)
                body_rows: list[list[str]] = []

                index += 2
                while index < len(lines):
                    body_line = lines[index]
                    body_stripped = body_line.strip()
                    if not body_stripped:
                        break

                    parsed_cells = _split_table_row(body_line)
                    if not parsed_cells:
                        break

                    body_rows.append(_normalize_table_row_cells(parsed_cells, expected_columns))
                    index += 1

                blocks.append(_render_table_html(normalized_header, body_rows))
                continue

        heading_match = MARKDOWN_HEADING_PATTERN.match(stripped)
        if heading_match:
            flush_paragraph()
            flush_list()
            level = len(heading_match.group(1))
            heading_text = render_inline_markdown(heading_match.group(2), escape_html=False)
            blocks.append(f"<h{level}>{heading_text}</h{level}>")
            index += 1
            continue

        ordered_match = MARKDOWN_ORDERED_LIST_PATTERN.match(stripped)
        if ordered_match:
            flush_paragraph()
            if list_type != "ol":
                flush_list()
                list_type = "ol"
                list_items = []
            list_items.append(render_inline_markdown(ordered_match.group(1), escape_html=False))
            index += 1
            continue

        unordered_match = MARKDOWN_UNORDERED_LIST_PATTERN.match(stripped)
        if unordered_match:
            flush_paragraph()
            if list_type != "ul":
                flush_list()
                list_type = "ul"
                list_items = []
            list_items.append(render_inline_markdown(unordered_match.group(1), escape_html=False))
            index += 1
            continue

        flush_list()
        paragraph_lines.append(line.strip())
        index += 1

    if in_code_fence:
        flush_code_fence()
    flush_paragraph()
    flush_list()

    return "\n".join(blocks)


def format_block(value: Any) -> Markup:
    """Escape plain text and convert it to safe HTML with line breaks.

    Applies confidence-token highlighting (CRITICAL, HIGH, etc.)
    and replaces newline characters with ``<br>`` tags.  Intended for
    use as a Jinja2 template filter.

    Args:
        value: Raw text to format.

    Returns:
        A :class:`~markupsafe.Markup` string safe for Jinja2
        rendering, or an N/A placeholder when *value* is empty.
    """
    text = _stringify(value, default="")
    if not text:
        return Markup('<span class="empty-value">N/A</span>')

    escaped = str(escape(text.replace("\r\n", "\n").replace("\r", "\n")))
    highlighted = highlight_confidence_tokens(escaped)
    with_line_breaks = highlighted.replace("\n", "<br>\n")
    return Markup(with_line_breaks)


def format_markdown_block(value: Any) -> Markup:
    """Convert Markdown text to HTML via :func:`markdown_to_html`.

    Intended for use as a Jinja2 template filter.

    Args:
        value: Raw Markdown text to render.

    Returns:
        A :class:`~markupsafe.Markup` string of rendered HTML, or
        an N/A placeholder when *value* is empty.
    """
    text = _stringify(value, default="")
    if not text:
        return Markup('<span class="empty-value">N/A</span>')
    return Markup(markdown_to_html(text))

File: static/js/markdown.js
/**
 * Markdown-to-DOM rendering for AIFT.
 *
 * Converts a subset of Markdown (headings, lists, tables, code fences,
 * inline bold/italic/code, and confidence tokens) into a DocumentFragment.
 *
 * Depends on: AIFT (utils.js)
 */
"use strict";

(() => {
  const A = window.AIFT;

  /**
   * Render a Markdown string into a DOM container, replacing its contents.
   *
   * @param {HTMLElement|null} container - Target DOM element.
   * @param {string} text - Markdown source text.
   * @param {string} emptyText - Placeholder shown when text is blank.
   */
  function renderMarkdownInto(container, text, emptyText) {
    if (!container) return;
    container.innerHTML = "";
    const raw = String(text || "");
    if (!raw.trim()) {
      const p = document.createElement("p");
      p.textContent = emptyText;
      container.appendChild(p);
      return;
    }
    container.appendChild(markdownToFragment(raw));
  }

  /**
   * Convert a Markdown string into a DocumentFragment.
   *
   * Supports headings, ordered/unordered lists, fenced code blocks,
   * tables, and inline formatting (bold, italic, code, confidence tokens).
   *
   * @param {string} text - Markdown source text.
   * @returns {DocumentFragment} Fragment ready for insertion into the DOM.
   */
  function markdownToFragment(text) {
    const fragment = document.createDocumentFragment();
    const lines = String(text || "").replace(/\r\n?/g, "\n").split("\n");
    let paragraphLines = [];
    let listNode = null;
    let listType = "";
    let inCodeFence = false;
    let codeFenceLines = [];

    /** Flush and append the current list node to the fragment. */
    const closeList = () => {
      if (!listNode) return;
      fragment.appendChild(listNode);
      listNode = null;
      listType = "";
    };

    /** Flush accumulated paragraph lines into a <p> element. */
    const flushParagraph = () => {
      if (!paragraphLines.length) return;
      const p = document.createElement("p");
      const html = renderInlineMarkdown(paragraphLines.join("\n")).replace(/\n/g, "<br>");
      p.innerHTML = html;
      fragment.appendChild(p);
      paragraphLines = [];
    };

    /** Flush accumulated code-fence lines into a <pre><code> block. */
    const flushCodeFence = () => {
      const pre = document.createElement("pre");
      const code = document.createElement("code");
      code.textContent = codeFenceLines.join("\n");
      pre.appendChild(code);
      fragment.appendChild(pre);
      codeFenceLines = [];
    };

    for (let index = 0; index < lines.length; index += 1) {
      const line = lines[index];
      const trimmed = String(line || "").trim();

      if (inCodeFence) {
        if (trimmed.startsWith("```")) {
          inCodeFence = false;
          flushCodeFence();
          continue;
        }
        codeFenceLines.push(line);
        continue;
      }

      if (trimmed.startsWith("```")) {
        flushParagraph();
        closeList();
        inCodeFence = true;
        codeFenceLines = [];
        continue;
      }

      if (!trimmed) {
        flushParagraph();
        closeList();
        continue;
      }

      const headerCells = splitMarkdownTableRow(line);
      if (headerCells && index + 1 < lines.length) {
        const separatorCells = splitMarkdownTableRow(lines[index + 1]);
        if (
          separatorCells
          && headerCells.length === separatorCells.length
          && isMarkdownTableSeparatorRow(separatorCells)
        ) {
          flushParagraph();
          closeList();

          const expectedColumns = headerCells.length;
          const normalizedHeader = normalizeMarkdownTableCells(headerCells, expectedColumns);

          const table = document.createElement("table");
          const thead = document.createElement("thead");
          const headerRow = document.createElement("tr");
          normalizedHeader.forEach((cell) => {
            const th = document.createElement("th");
            th.innerHTML = renderInlineMarkdown(cell);
            headerRow.appendChild(th);
          });
          thead.appendChild(headerRow);
          table.appendChild(thead);

          const tbody = document.createElement("tbody");
          let hasBodyRows = false;
          index += 2;
          while (index < lines.length) {
            const bodyLine = lines[index];
            const bodyTrimmed = String(bodyLine || "").trim();
            if (!bodyTrimmed) break;

            const parsedBodyCells = splitMarkdownTableRow(bodyLine);
            if (!parsedBodyCells) break;

            const normalizedBody = normalizeMarkdownTableCells(parsedBodyCells, expectedColumns);
            const tr = document.createElement("tr");
            normalizedBody.forEach((cell) => {
              const td = document.createElement("td");
              td.innerHTML = renderInlineMarkdown(cell);
              tr.appendChild(td);
            });
            tbody.appendChild(tr);
            hasBodyRows = true;
            index += 1;
          }

          if (hasBodyRows) table.appendChild(tbody);
          fragment.appendChild(table);
          index -= 1;
          continue;
        }
      }

      const heading = trimmed.match(/^(#{1,6})\s+(.*)$/);
      if (heading) {
        flushParagraph();
        closeList();
        const level = heading[1].length;
        const content = heading[2] || "";
        const h = document.createElement(`h${level}`);
        h.innerHTML = renderInlineMarkdown(content);
        fragment.appendChild(h);
        continue;
      }

      const ordered = trimmed.match(/^\d+\.\s+(.*)$/);
      if (ordered) {
        flushParagraph();
        if (listType !== "ol") {
          closeList();
          listNode = document.createElement("ol");
          listType = "ol";
        }
        const li = document.createElement("li");
        li.innerHTML = renderInlineMarkdown(ordered[1] || "");
        listNode.appendChild(li);
        continue;
      }

      const unordered = trimmed.match(/^[-*]\s+(.*)$/);
      if (unordered) {
        flushParagraph();
        if (listType !== "ul") {
          closeList();
          listNode = document.createElement("ul");
          listType = "ul";
        }
        const li = document.createElement("li");
        li.innerHTML = renderInlineMarkdown(unordered[1] || "");
        listNode.appendChild(li);
        continue;
      }

      closeList();
      paragraphLines.push(trimmed);
    }

    if (inCodeFence) flushCodeFence();
    flushParagraph();
    closeList();
    return fragment;
  }

  /**
   * Split a Markdown table row into trimmed cell strings.
   *
   * @param {string} line - A single line of text.
   * @returns {string[]|null} Array of cell values, or null if not a table row.
   */
  function splitMarkdownTableRow(line) {
    const raw = String(line || "");
    if (!raw.includes("|")) return null;
    let trimmed = raw.trim();
    if (!trimmed || !trimmed.includes("|")) return null;
    if (trimmed.startsWith("|")) trimmed = trimmed.slice(1);
    if (trimmed.endsWith("|")) trimmed = trimmed.slice(0, -1);
    return trimmed.split("|").map((cell) => cell.trim());
  }

  /**
   * Check whether an array of cells forms a valid Markdown table separator
   * row (e.g. `| --- | :---: |`).
   *
   * @param {string[]} cells - Cell values from splitMarkdownTableRow.
   * @returns {boolean}
   */
  function isMarkdownTableSeparatorRow(cells) {
    if (!Array.isArray(cells) || !cells.length) return false;
    return cells.every((cell) => /^:?-{3,}:?$/.test(String(cell || "").trim()));
  }

  /**
   * Pad or trim an array of table cells to exactly `expectedCount` entries.
   *
   * @param {string[]} cells - Raw cells array.
   * @param {number} expectedCount - Target column count.
   * @returns {string[]} Normalised array of trimmed cell strings.
   */
  function normalizeMarkdownTableCells(cells, expectedCount) {
    const normalized = Array.isArray(cells)
      ? cells.slice(0, expectedCount).map((cell) => String(cell || "").trim())
      : [];
    while (normalized.length < expectedCount) normalized.push("");
    return normalized;
  }

  /**
   * Render inline Markdown (bold, italic, code, confidence tokens) to HTML.
   *
   * Code spans are extracted first so their content is not processed for
   * bold/italic, then the remaining text is HTML-escaped and formatted.
   * The guards mirror the Python renderer (app/reporter/markdown.py): a
   * code span needs both delimiting backticks (a lone backtick stays
   * literal), and star italics use lookarounds so stray `*` characters next
   * to `**` are not treated as emphasis markers.
   *
   * @param {string} text - Inline Markdown source.
   * @returns {string} HTML string with inline formatting applied.
   */
  function renderInlineMarkdown(text) {
    const source = String(text || "");
    if (!source) return "";
    const parts = source.split(/(`[^`\n]*`)/g);
    return parts
      .map((part) => {
        if (part.length >= 2 && part.startsWith("`") && part.endsWith("`")) {
          return `<code>${A.escapeHtml(part.slice(1, -1))}</code>`;
        }
        let out = A.escapeHtml(part);
        out = out.replace(/\*\*(.+?)\*\*/g, "<strong>$1</strong>");
        out = out.replace(/(^|\s)__([^\s_](?:.*?[^\s_])?)__(?=$|\s|[.,;!?:)])/g, "$1<strong>$2</strong>");
        out = out.replace(/(?<!\*)\*(?!\*)(.+?)(?<!\*)\*(?!\*)/g, "<em>$1</em>");
        out = out.replace(/(^|\s)_([^\s_](?:.*?[^\s_])?)_(?=$|\s|[.,;!?:)])/g, "$1<em>$2</em>");
        out = highlightConfidenceTokens(out);
        return out;
      })
      .join("");
  }

  /**
   * Wrap confidence-level tokens in explicit confidence contexts in coloured
   * `<span>` badges using CONFIDENCE_CLASS_MAP.
   *
   * @param {string} text - HTML string (already escaped).
   * @returns {string} HTML with confidence tokens highlighted.
   */
  function highlightConfidenceTokens(text) {
    A.CONFIDENCE_TOKEN_PATTERN.lastIndex = 0;
    return String(text || "").replace(A.CONFIDENCE_TOKEN_PATTERN, (match, token) => {
      const normalized = String(token || match || "").toUpperCase();
      const cssClass = A.CONFIDENCE_CLASS_MAP[normalized] || "confidence-unknown";
      const tokenIndex = match.toLowerCase().lastIndexOf(String(token || "").toLowerCase());
      const prefix = tokenIndex >= 0 ? match.slice(0, tokenIndex) : "";
      const suffix = tokenIndex >= 0 ? match.slice(tokenIndex + String(token || "").length) : "";
      return `${prefix}<span class="confidence-inline ${cssClass}">${normalized}</span>${suffix}`;
    });
  }

  // ── Public API ─────────────────────────────────────────────────────────────
  A.renderMarkdownInto = renderMarkdownInto;
})();
